daemon/containerd/cache.go

```diff
@@ -2,211 +2,238 @@ package containerd
+	if err != nil {
+		return nil, fmt.Errorf("GetImage: %w", err)
+	}
 
-	// FROM scratch
-	if parentID == "" {
-		c, err := ic.imageService.getImagesWithLabel(ctx, imageLabelClassicBuilderFromScratch, "1")
-		if err != nil {
-			return "", err
-		}
-		children = c
-	} else {
-		c, err := ic.imageService.Children(ctx, image.ID(parentID))
-		if err != nil {
-			return "", err
-		}
-		children = c
+	c8dImg, err := c.is.resolveImage(ctx, ref)
+	if err != nil {
+		return nil, fmt.Errorf("resolveImage: %w", err)
 	}
 
-	var match *image.Image
-	for _, child := range children {
-		ccDigestStr, err := ic.imageService.getImageLabelByDigest(ctx, child.Digest(), imageLabelClassicBuilderContainerConfig)
+	var errFound = errors.New("success")
+	err = c.is.walkImageManifests(ctx, c8dImg, func(img *ImageManifest) error {
+		desc, err := img.Config(ctx)
 		if err != nil {
-			return "", err
-		}
-		if ccDigestStr == "" {
-			continue
+			log.G(ctx).WithFields(log.Fields{
+				"image": img,
+				"error": err,
+			}).Warn("failed to get config descriptor for image")
+			return nil
 		}
 
-		dgst, err := digest.Parse(ccDigestStr)
+		info, err := c.is.content.Info(ctx, desc.Digest)
 		if err != nil {
-			log.G(ctx).WithError(err).Warnf("invalid container config digest: %q", ccDigestStr)
-			continue
-		}
-
-		var cc container.Config
-		if err := readConfig(ctx, ic.imageService.content, ocispec.Descriptor{Digest: dgst}, &cc); err != nil {
-			if errdefs.IsNotFound(err) {
-				log.G(ctx).WithError(err).WithField("image", child).Warnf("missing container config: %q", ccDigestStr)
-				continue
+			if !cerrdefs.IsNotFound(err) {
+				log.G(ctx).WithFields(log.Fields{
+					"image": img,
+					"desc":  desc,
+					"error": err,
+				}).Warn("failed to get info of image config")
 			}
-			return "", err
+			return nil
 		}
 
-		if cache.CompareConfig(&cc, cfg) {
-			childImage, err := ic.imageService.GetImage(ctx, child.String(), backend.GetImageOpts{Platform: &platform})
+		if dgstStr, ok := info.Labels[contentLabelGcRefContainerConfig]; ok {
+			dgst, err := digest.Parse(dgstStr)
 			if err != nil {
-				if errdefs.IsNotFound(err) {
-					continue
-				}
-				return "", err
+				log.G(ctx).WithFields(log.Fields{
+					"label":   contentLabelClassicBuilderImage,
+					"value":   dgstStr,
+					"content": desc.Digest,
+					"error":   err,
+				}).Warn("invalid digest in label")
+				return nil
+			}
+
+			configDesc := ocispec.Descriptor{
+				Digest: dgst,
 			}
 
-			if childImage.Created != nil && (match == nil || match.Created.Before(*childImage.Created)) {
-				match = childImage
+			var config container.Config
+			if err := readConfig(ctx, c.is.content, configDesc, &config); err != nil {
+				if !errdefs.IsNotFound(err) {
+					log.G(ctx).WithFields(log.Fields{
+						"configDigest": dgst,
+						"error":        err,
+					}).Warn("failed to read container config")
+				}
+				return nil
 			}
+
+			outImg.ContainerConfig = config
+			return errFound
```

Comment:
Trying to understand this bit, are we doing this so we can stop as soon as we've found the image we want? Wonder if we should add a comment.

Reply:
Yes, no need to read other images if we already found the one we're interested in. Can add a comment.

Reply:
please 🙏

Reply:
Let me know if you have some more comments, so I can batch them up in one force push 😄

Reply:
Done, thanks!